public/app/features/dashboard-scene/scene/DashboardScene.tsx

```diff
@@ -369,40 +383,61 @@ export class DashboardScene extends SceneObjectBase<DashboardSceneState> {
       (event: SceneObjectStateChangedEvent) => {
         if (event.payload.changedObject instanceof SceneRefreshPicker) {
           if (Object.prototype.hasOwnProperty.call(event.payload.partialUpdate, 'intervals')) {
-            this.setIsDirty();
+            this.postChangesMessage();
           }
         }
         if (event.payload.changedObject instanceof SceneDataLayers) {
-          this.setIsDirty();
+          this.postChangesMessage();
         }
         if (event.payload.changedObject instanceof dataLayers.AnnotationsDataLayer) {
           if (!Object.prototype.hasOwnProperty.call(event.payload.partialUpdate, 'data')) {
-            this.setIsDirty();
+            this.postChangesMessage();
           }
         }
         if (event.payload.changedObject instanceof SceneGridItem) {
-          this.setIsDirty();
+          this.postChangesMessage();
         }
         if (event.payload.changedObject instanceof DashboardScene) {
           if (Object.keys(event.payload.partialUpdate).some((key) => PERSISTED_PROPS.includes(key))) {
-            this.setIsDirty();
+            this.postChangesMessage();
           }
         }
         if (event.payload.changedObject instanceof SceneTimeRange) {
-          this.setIsDirty();
+          this.postChangesMessage();
         }
         if (event.payload.changedObject instanceof DashboardControls) {
           if (Object.prototype.hasOwnProperty.call(event.payload.partialUpdate, 'hideTimeControls')) {
-            this.setIsDirty();
+            this.postChangesMessage();
           }
         }
+        if (event.payload.changedObject instanceof SceneVariableSet) {
+          this.postChangesMessage();
+        }
+        if (event.payload.changedObject instanceof DashboardAnnotationsDataLayer) {
+          this.postChangesMessage();
+        }
+        if (isSceneVariableInstance(event.payload.changedObject)) {
+          this.postChangesMessage();
+        }
       }
     );
   }
 
-  private setIsDirty() {
-    if (!this.state.isDirty) {
-      this.setState({ isDirty: true });
+  private postChangesMessage() {
+    this._changesWorker?.postMessage({ changed: transformSceneToSaveModel(this), initial: this._initialSaveModel });
+  }
+
+  private detectChanges(changes: DashboardChangeInfo) {
+    const { hasChanges } = changes;
+
+    if (hasChanges) {
+      if (!this.state.isDirty) {
+        this.setState({ isDirty: true });
+      }
+    } else {
+      if (this.state.isDirty) {
+        this.setState({ isDirty: false });
+      }
     }
```

Comment:
would like to move this, the _changesWorker  and the state subscription with all the "changeObject instanceof xx" to a separate DashboardSceneChangeTracker. That way we can have a class that only handles and cares about change tracking 

I don't think we need the workers folder, maybe we move all this to saving (getDashboardChanges, getDashboardChangesForScene, worker, etc)? since it's all related to saving).

Reply:
I moved everything and refactored it a bit. How does it look now? :)